Fix mac deployment target for building with XCode 8+

STRML opened the pull request:
Fixes "error: 'promise<bool>' is unavailable: introduced in macOS 10.8" etc.

Required for building on updated Macs.

goatpig commented:
Already had this discussion with @droark, we're not going to discontinue OSX 10.7 support until the new wallets and signer are out (so that there is at least one offline signer of the new code build for 10.7). 

If you want to support XCode 8 now, please rework your PR to only drop OSX 10.7 support when using XCode 8.

Also, please avoid submitting a PR with 90% white space changes.

STRML commented:
Sorry, this didn't come up on a search. Will amend.

droark commented:
FWIW, you could rewrite the PR to have the Makefile use *xcodebuild -version* to determine which version of XCode is present. If 8, jump to 10.8, otherwise leave at 10.7. Not ideal but it seems to be the only reasonable solution.

STRML commented:
Yep, that's what I was doing - updated the commit.

droark commented:
utACK

Slight nit: I don't know if it's an XCode bug, per se. [This comment](https://github.com/bitcoin/bitcoin/issues/8577#issuecomment-255945996) describes what's going on. I suppose the message is fine as is, especially since it'll get chopped once we drop 10.7.

STRML commented:
Hmm - so features that were buggy in 10.7 were simply marked as removed in XCode 8? I get about 20 of these without this PR:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:632:1: error: '__assoc_state<std::__1::shared_ptr<Payload> >' is unavailable: introduced in macOS 10.8
__assoc_state<_Rp>::__on_zero_shared() _NOEXCEPT
^
...
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1422:1: error: 'promise<bool>' is unavailable: introduced in macOS 10.8
promise<_Rp>::promise()
^
```

droark commented:
@STRML - I think it's more that the library is buggy and/or incomplete in 10.7, and XCode 8 just refuses to build towards it for that reason. I presume that Apple's stance is that it's cool if the code works on 10.7. They just won't guarantee that it works properly, whereas 10.8 is complete and good to go. I could be wrong, though.

droark commented on cppForSwig/Makefile:
```diff
@@ -17,7 +17,14 @@ endif
 ifeq ($(shell uname), Darwin)
+# XCode 8 breaks when compiling for 10.7, XCode 7 and below does not
+XCODE_VERSION:=$(shell xcodebuild -version | awk '{ print $$2 }' | head -n1)
+ifeq ($(shell echo "${XCODE_VERSION} < 8" | bc), 0)
```
This line ought to be version > 7. In addition, I'm getting the following error on 10.12.3.

```(standard_in) 1: parse error```

STRML replied:
Actually, it's fine as written - notice it's comparing to `0`. That's obviously a bit confusing though so I've switched it.

As for the parse error, I couldn't believe I would have shipped this broken then realized the latest XCode version is `8.2.1` (at the time it was `8.1`) - `8.2.1` is not a valid float and thus `bc` calls it an error.

droark commented:
@STRML - Can you double check your code? Finally tested it now that dev is open for fixes again. This doesn't work for me.

goatpig commented:
What is the status of this PR?

droark commented:
@goatpig - I said utACK awhile back but a test shows this isn't ready. I'm waiting for @STRML to fix it. If nothing happens in the next week or two, I'll just do another PR and ask that this one be closed.

STRML commented:
Amended; error was due to the latest XCode version (`8.2.1`) not being a valid float, whereas at the time of writing it was (`8.1`) - that's obviously brittle so I `cut` the version now.

droark commented on cppForSwig/Makefile:
```diff
@@ -17,7 +17,14 @@ endif
+# XCode 8 breaks when compiling for 10.7, XCode 7 and below does not
+XCODE_VERSION:=$(shell xcodebuild -version | awk '{ print $$2 }' | head -n1 | cut -d "." -f 1)
+ifeq ($(shell echo "${XCODE_VERSION} >= 8" | bc), 1)
+$(info XCode 8 Detected. Due to an XCode bug, only 10.8+ targets are supported.)
```
Okay. This works! I made a couple of changes that I'd like to recommend a couple of changes. I've pasted the code below.

```
# XCode 8 breaks when compiling for 10.7, XCode 7 and below does not
XCODE_VERSION_FULL:=$(shell xcodebuild -version | awk '{ print $$2 }')
XCODE_VERSION:=$(shell xcodebuild -version | awk '{ print $$2 }' | head -n1 | cut -d "." -f 1)
$(info XCode ${XCODE_VERSION_FULL} detected.)
ifeq ($(shell echo "${XCODE_VERSION} >= 8" | bc), 1)
$(info Due to XCode 8+ changes, the deployment target is 10.8. Use XCode 7 if a target of 10.7 is needed.)
MACOSX_DEPLOYMENT_TARGET=10.8
else
$(info A version of XCode below XCode 8 has been detected. The deployment target is 10.7.)
MACOSX_DEPLOYMENT_TARGET=10.7
endif
```

droark commented:
@STRML - Thanks. Would like to request a minor change but it's a nit more than anything else. Let us know if you'd like to incorporate the request.

@goatpig - Once @STRML changes things (or says he won't), this will be ready to go.

STRML commented:
That works for me @droark - I like verbose - already amended.

droark commented:
@STRML - Thanks!

@goatpig - ACK
